### PR TITLE
Update go 1.21 and fix lint issues

### PR DESCRIPTION
**Pull request by mpanchajanya**
### What this PR does / why we need it

 -  Updated github workflows to use go 1.21
 -  Updated root and test go.mod to use go v1.21
 -  Update tools like linters to latest
 -  Fix lint issues
 -  Removed deprecated linters deadcode, varcheck, structcheck, unparam
 
### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

 - Build the Tanzu CLI
 - Perform basic lifecycle ops
 - Unit tests, Lint checks
 
### Release note

```release-note
Tanzu CLI is updated to latest go v1.21
```

### CODE REVIEW

File: .golangci.yaml [hunk 1]
```diff
@@ -75,14 +131,10 @@ linters:
...
     - unconvert
     - unused
-    - varcheck
     - whitespace
     - bodyclose
     - noctx
     - rowserrcheck
-    - structcheck
-    - unparam
-
```
Comment (anujc25) [inline .golangci.yaml:85]: Are these removals intentional?

Comment (mpanchajanya) [reply .golangci.yaml:85]: Yes, these are deprecated, updated the PR description

Comment (mpanchajanya) [reply .golangci.yaml:85]: Few of the linters need to address separately as many files were flaged with lint issues.
Added TODO in the golangci.yaml

**Review by anujc25 — APPROVED** [review]
LGTM. Thanks for all the changes to use go v1.21
Let's file an issue to address the linter that needs to be enabled after addressing the errors.

**Comment by mpanchajanya** [comment]
> LGTM. Thanks for all the changes to use go v1.21 Let's file an issue to address the linter that needs to be enabled after addressing the errors.

Created JIRA to fix the disbaled lint issues

**Pull request merged**